packages/dev/core/src/Engines/nativeEngine.ts

```diff
@@ -722,20 +714,21 @@ export class NativeEngine extends Engine {
             this.onAfterShaderCompilationObservable.notifyObservers(this);
         };
 
-        if (this.isAsync(pipelineContext)) {
-            return this._engine.createProgramAsync(vertexCode, fragmentCode, onSuccess, (error: Error) => {
+        if (pipelineContext.isAsync) {
+            nativePipelineContext.program = this._engine.createProgramAsync(vertexCode, fragmentCode, onSuccess, (error: Error) => {
                 nativePipelineContext.compilationError = error;
-            }) as WebGLProgram;
+            });
         } else {
             try {
-                const program = (nativePipelineContext.nativeProgram = this._engine.createProgram(vertexCode, fragmentCode));
+                nativePipelineContext.program = this._engine.createProgram(vertexCode, fragmentCode);
                 onSuccess();
-                return program as WebGLProgram;
-            } catch (e: any) {
+            } catch (e) {
                 const message = e?.message;
                 throw new Error("SHADER ERROR" + (typeof message === "string" ? "\n" + message : ""));
             }
         }
+
+        return nativePipelineContext.program as WebGLProgram;

```

Comment:
I don't think `program` is really a `WebGLProgram` in Native, so casting is only to comply with the signature?

I think you should do as in WebGPU, throw a "Not available on Native" error from `createShaderProgram` / `createRawShaderProgram` and use other names for these (internal) methods. These methods can't be used outside the WebGL engine.

Reply:
This was done a while ago, so yes, it's just for typing. We can do the same as WebGPU like you say, but I would rather do that separately.